File: packages/ts-host/src/store-backend/chat-store-backend.ts
import { currentBearer } from '../als'
import type {
  EditResult,
  FileInfo,
  GlobResult,
  GrepMatch,
  GrepResult,
  LsResult,
  ReadRawResult,
  ReadResult,
  StoreBackend,
  WriteResult,
} from './types'

/** DTO REST истории chat-backend. */
interface ProjectDto {
  id: string
  slug: string | null
  title: string
  createdAt: string
  updatedAt: string
}
interface ThreadDto {
  id: string
  contextId: string
  slug: string | null
  title: string | null
  projectId: string | null
  createdAt: string
  updatedAt: string
}
interface ChatMessageDto {
  role: string
  content: string
  a2uiArtifacts: unknown
  createdAt: string
}
/** Серверное FTS-совпадение (grep). */
interface ChatGrepHit {
  contextId: string
  threadSlug: string | null
  threadTitle: string | null
  projectSlug: string | null
  snippet: string
}

export interface ChatStoreBackendOptions {
  /**
   * База REST chat-backend (history API). Backend ходит в вложенные ресурсы:
   * `/api/projects/`, `/api/projects/:slug/`, `/api/projects/:slug/threads/[:ts]`, `/api/threads/`,
   * `/api/threads/:slug`, с серверным поиском через `?name=` (glob) и `?content=` (grep, FTS).
   */
  baseUrl: string
  /** user-JWT для форварда (приватность per-user). По умолчанию `currentBearer` из request-scope. */
  bearer?: () => string | undefined
  /** fetch (по умолчанию глобальный). */
  fetchImpl?: typeof fetch
}

const READ_ONLY = 'ChatStoreBackend доступен только для чтения (история чатов/проектов)'

/**
 * StoreBackend поверх history-API chat-backend: read-only доступ агентов к истории чатов и проектам
 * через файловую абстракцию deepagents `CompositeBackend`. ВСЕ операции серверные (включая поиск) —
 * backend лишь маппит virtual-path + query на REST endpoint и форматирует ответ. Чат — основная
 * сущность, проект — опциональная группировка.
 *
 * Виртуальная ФС (якоря `projects`/`threads` в пути):
 * - `/projects/`                          — проекты;
 * - `/projects/{slug}/`                   — проект (его чаты);
 * - `/projects/{slug}/threads/`           — чаты проекта;
 * - `/projects/{slug}/threads/{ts}/`      — чат в проекте;
 * - `/threads/`                           — чаты вне проектов;
 * - `/threads/{ts}/`                      — чат вне проекта.
 *
 * Операции:
 * - `ls`   — структурный листинг (FileInfo);
 * - `read` — markdown: индекс проектов/чатов или расшифровка переписки; offset/limit режут строки;
 * - `glob(pat, path)` — серверный поиск по ИМЕНИ (`?name=`, ILIKE) → FileInfo[];
 * - `grep(pat, path)` — серверный full-text по СОДЕРЖИМОМУ (`?content=`, Postgres-FTS) → GrepMatch[];
 * - `write/edit` — read-only (ошибка).
 *
 * Монтирование (сохраняет якоря после обрезки префикса CompositeBackend):
 * ```ts
 * new CompositeBackend(new StateBackend(), {
 *   "/history/": new ChatStoreBackend({ baseUrl: process.env.CHAT_BACKEND_URL! }),
 * })
 * // агент: ls('/history/projects/'), read('/history/projects/{slug}/threads/{ts}/'),
 * //        grep('лифты','/history/projects/'), glob('расч','/history/threads/')
 * ```
 */
export class ChatStoreBackend implements StoreBackend {
  private readonly baseUrl: string
  private readonly bearer: () => string | undefined
  private readonly fetchImpl: typeof fetch

  constructor(opts: ChatStoreBackendOptions) {
    this.baseUrl = opts.baseUrl.replace(/\/+$/, '')
    this.bearer = opts.bearer ?? currentBearer
    this.fetchImpl = opts.fetchImpl ?? globalThis.fetch
  }

  // ── ls ─────────────────────────────────────────────────────────────────────
  async ls(path: string): Promise<LsResult> {
    try {
      const p = parsePath(path)
      switch (p.kind) {
        case 'root':
          return { files: [dir('/projects/'), dir('/threads/')] }
        case 'projects': {
          const { projects } = await this.api<{ projects: ProjectDto[] }>('/api/projects/')
          return { files: projects.map((pr) => dir(projectDir(pr))) }
        }
        case 'project':
          // содержимое проекта = поддиректория threads/
          return { files: [dir(`/projects/${p.projectSlug}/threads/`)] }
        case 'project-threads': {
          const { threads } = await this.api<{ threads: ThreadDto[] }>(
            `/api/projects/${enc(p.projectSlug)}/threads/`,
          )
          return { files: threads.map((t) => file(`/projects/${p.projectSlug}/threads/${threadSeg(t)}`, t)) }
        }
        case 'threads': {
          const { threads } = await this.api<{ threads: ThreadDto[] }>('/api/threads/?projectId=none')
          return { files: threads.map((t) => file(`/threads/${threadSeg(t)}`, t)) }
        }
        default:
          return { error: `Не директория: ${path}` }
      }
    } catch (e) {
      return { error: errMsg(e) }
    }
  }

  // ── read ───────────────────────────────────────────────────────────────────
  async read(path: string, offset?: number, limit?: number): Promise<ReadResult> {
    try {
      const md = await this.renderRead(parsePath(path), path)
      if (md === null) return { error: `Не найдено: ${path}` }
      return { content: sliceLines(md, offset, limit), mimeType: 'text/markdown' }
    } catch (e) {
      return { error: errMsg(e) }
    }
  }

  // ── glob (по имени, серверный ILIKE) ─────────────────────────────────────────
  async glob(pattern: string, path?: string): Promise<GlobResult> {
    try {
      const name = stripGlob(pattern)
      const p = parsePath(path ?? '/')
      if (p.kind === 'root' || p.kind === 'projects') {
        const { projects } = await this.api<{ projects: ProjectDto[] }>('/api/projects/', { name })
        return { files: projects.map((pr) => dir(projectDir(pr))) }
      }
      if (p.kind === 'project' || p.kind === 'project-threads') {
        const { threads } = await this.api<{ threads: ThreadDto[] }>(
          `/api/projects/${enc(p.projectSlug)}/threads/`,
          { name },
        )
        return { files: threads.map((t) => file(`/projects/${p.projectSlug}/threads/${threadSeg(t)}`, t)) }
      }
      if (p.kind === 'threads') {
        const { threads } = await this.api<{ threads: ThreadDto[] }>('/api/threads/', {
          name,
          projectId: 'none',
        })
        return { files: threads.map((t) => file(`/threads/${threadSeg(t)}`, t)) }
      }
      return { files: [] }
    } catch (e) {
      return { error: errMsg(e) }
    }
  }

  // ── grep (по содержимому, серверный FTS) ─────────────────────────────────────
  async grep(pattern: string, path?: string | null): Promise<GrepResult> {
    try {
      const p = parsePath(path ?? '/')
      let hits: ChatGrepHit[]
      if (p.kind === 'project' || p.kind === 'project-threads') {
        hits = (
          await this.api<{ matches: ChatGrepHit[] }>(
            `/api/projects/${enc(p.projectSlug)}/threads/`,
            { content: pattern },
          )
        ).matches
      } else if (p.kind === 'threads') {
        hits = (
          await this.api<{ matches: ChatGrepHit[] }>('/api/threads/', {
            content: pattern,
            projectId: 'none',
          })
        ).matches
      } else {
        // root/projects/thread — поиск по всем чатам пользователя
        hits = (await this.api<{ matches: ChatGrepHit[] }>('/api/projects/', { content: pattern }))
          .matches
      }
      return {
        matches: hits.map<GrepMatch>((h) => ({
          path: grepHitPath(h),
          line: 1,
          text: `[${h.threadTitle ?? 'чат'}] ${oneLine(h.snippet)}`,
        })),
      }
    } catch (e) {
      return { error: errMsg(e) }
    }
  }

  // ── read-only ────────────────────────────────────────────────────────────────
  write(): Promise<WriteResult> {
    return Promise.resolve({ error: READ_ONLY })
  }
  edit(): Promise<EditResult> {
    return Promise.resolve({ error: READ_ONLY })
  }
  // История — markdown-текст (read с пагинацией); raw-бинарь не поддерживаем.
  readRaw(): Promise<ReadRawResult> {
    return Promise.resolve({ error: READ_ONLY })
  }

  // ── рендереры read ────────────────────────────────────────────────────────────
  private async renderRead(p: Parsed, raw: string): Promise<string | null> {
    switch (p.kind) {
      case 'root': {
        const { projects } = await this.api<{ projects: ProjectDto[] }>('/api/projects/')
        const lines = ['# История', '', '## Проекты']
        for (const pr of projects) lines.push(`- ${pr.title} — \`/projects/${projSeg(pr)}/\``)
        if (projects.length === 0) lines.push('_нет проектов_')
        lines.push('', '## Чаты вне проектов — `/threads/`')
        return lines.join('\n')
      }
      case 'projects': {
        const { projects } = await this.api<{ projects: ProjectDto[] }>('/api/projects/')
        const lines = ['# Проекты', '']
        for (const pr of projects) lines.push(`- ${pr.title} — \`/projects/${projSeg(pr)}/\``)
        if (projects.length === 0) lines.push('_нет проектов_')
        return lines.join('\n')
      }
      case 'project': {
        const { project } = await this.api<{ project: ProjectDto }>(
          `/api/projects/${enc(p.projectSlug)}/`,
        )
        const { threads } = await this.api<{ threads: ThreadDto[] }>(
          `/api/projects/${enc(p.projectSlug)}/threads/`,
        )
        return renderThreadList(`# Проект: ${project.title}`, threads, `/projects/${p.projectSlug}/threads`)
      }
      case 'project-threads': {
        const { threads } = await this.api<{ threads: ThreadDto[] }>(
          `/api/projects/${enc(p.projectSlug)}/threads/`,
        )
        return renderThreadList('# Чаты проекта', threads, `/projects/${p.projectSlug}/threads`)
      }
      case 'threads': {
        const { threads } = await this.api<{ threads: ThreadDto[] }>('/api/threads/?projectId=none')
        return renderThreadList('# Чаты вне проектов', threads, '/threads')
      }
      case 'project-thread': {
        const { messages } = await this.api<{ messages: ChatMessageDto[] }>(
          `/api/projects/${enc(p.projectSlug)}/threads/${enc(p.threadSlug)}`,
        )
        return renderChat(p.threadSlug, messages)
      }
      case 'thread': {
        const { messages } = await this.api<{ thread: ThreadDto; messages: ChatMessageDto[] }>(
          `/api/threads/${enc(p.threadSlug)}`,
        )
        return renderChat(p.threadSlug, messages)
      }
      default:
        return null
    }
    void raw
  }

  // ── HTTP ─────────────────────────────────────────────────────────────────────
  private async api<T>(path: string, query?: Record<string, string | undefined>): Promise<T> {
    const url = new URL(`${this.baseUrl}${path}`)
    if (query) {
      for (const [k, v] of Object.entries(query)) if (v !== undefined) url.searchParams.set(k, v)
    }
    const headers: Record<string, string> = { Accept: 'application/json' }
    const token = this.bearer()
    if (token) headers.Authorization = `Bearer ${token}`
    const res = await this.fetchImpl(url.toString(), { headers })
    if (!res.ok) throw new Error(`chat-backend ${path} → HTTP ${res.status}`)
    return res.json() as Promise<T>
  }
}

// ── path parsing (толерантно к scope-обрезке: ищем якоря в сегментах) ────────────
type Parsed =
  | { kind: 'root' }
  | { kind: 'projects' }
  | { kind: 'project'; projectSlug: string }
  | { kind: 'project-threads'; projectSlug: string }
  | { kind: 'project-thread'; projectSlug: string; threadSlug: string }
  | { kind: 'threads' }
  | { kind: 'thread'; threadSlug: string }
  | { kind: 'unknown' }

function parsePath(path: string): Parsed {
  const seg = path.split('/').filter(Boolean)
  if (seg.length === 0) return { kind: 'root' }
  const i = seg.findIndex((s) => s === 'projects' || s === 'threads')
  if (i === -1) return { kind: 'unknown' }
  if (seg[i] === 'projects') {
    const projectSlug = seg[i + 1]
    if (!projectSlug) return { kind: 'projects' }
    const sub = seg[i + 2]
    if (!sub) return { kind: 'project', projectSlug }
    if (sub === 'threads') {
      const threadSlug = seg[i + 3]
      return threadSlug
        ? { kind: 'project-thread', projectSlug, threadSlug }
        : { kind: 'project-threads', projectSlug }
    }
    return { kind: 'unknown' }
  }
  const threadSlug = seg[i + 1]
  return threadSlug ? { kind: 'thread', threadSlug } : { kind: 'threads' }
}

// ── helpers ─────────────────────────────────────────────────────────────────────
function dir(path: string): FileInfo {
  return { path, is_dir: true }
}
function file(path: string, t: ThreadDto): FileInfo {
  return { path, is_dir: false, modified_at: t.updatedAt }
}
function projSeg(p: ProjectDto): string {
  return p.slug ?? p.id
}
function projectDir(p: ProjectDto): string {
  return `/projects/${projSeg(p)}/`
}
function threadSeg(t: ThreadDto): string {
  return t.slug ?? t.contextId
}
function grepHitPath(h: ChatGrepHit): string {
  const ts = h.threadSlug ?? h.contextId
  return h.projectSlug ? `/projects/${h.projectSlug}/threads/${ts}` : `/threads/${ts}`
}
function renderThreadList(header: string, threads: ThreadDto[], base: string): string {
  const lines = [header, '']
  for (const t of threads) {
    lines.push(`- ${t.title ?? 'Без названия'} — \`${base}/${threadSeg(t)}\` (обновлён: ${t.updatedAt})`)
  }
  if (threads.length === 0) lines.push('_нет чатов_')
  return lines.join('\n')
}
function renderChat(slug: string, messages: ChatMessageDto[]): string {
  const lines = [`# Чат ${slug}`, '']
  for (const m of messages) {
    const role = m.role === 'user' ? 'Пользователь' : m.role === 'assistant' ? 'Ассистент' : m.role
    lines.push(`**${role}:** ${typeof m.content === 'string' ? m.content : ''}`, '')
  }
  if (messages.length === 0) lines.push('_пусто_')
  return lines.join('\n')
}
function enc(s: string): string {
  return encodeURIComponent(s)
}
function errMsg(e: unknown): string {
  return e instanceof Error ? e.message : String(e)
}
function oneLine(s: string): string {
  return s.replace(/\s+/g, ' ').trim().slice(0, 200)
}
/** glob-маска → подстрока для серверного ILIKE (сервер ищет contains, спецсимволы glob убираем). */
function stripGlob(pattern: string): string {
  return pattern.replace(/[*?]/g, '').trim()
}
function sliceLines(content: string, offset?: number, limit?: number): string {
  if (!offset && limit === undefined) return content
  const lines = content.split('\n')
  const start = offset ?? 0
  const end = limit === undefined ? lines.length : start + limit
  return lines.slice(start, end).join('\n')
}

File: packages/python-host/src/ai37_agent_host/store_backend/chat_store_backend.py
"""История чатов/проектов — порт ``ts-host/src/store-backend/chat-store-backend.ts``.

Read-only StoreBackend поверх history-API chat-backend: агент читает историю чатов и проекты
пользователя через файловую абстракцию deepagents ``CompositeBackend``. ВСЕ операции серверные
(включая поиск) — backend лишь маппит virtual-path + query на REST-endpoint и форматирует ответ.
Чат — основная сущность, проект — опциональная группировка.

Виртуальная ФС (якоря ``projects``/``threads`` в пути):
``/projects/``, ``/projects/{slug}/``, ``/projects/{slug}/threads/``,
``/projects/{slug}/threads/{ts}/``, ``/threads/``, ``/threads/{ts}/``.
"""

from __future__ import annotations

import re
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any
from urllib.parse import quote

import httpx

from ..als import current_bearer
from .types import (
    EditResult,
    FileInfo,
    GlobResult,
    GrepMatch,
    GrepResult,
    LsResult,
    ReadRawResult,
    ReadResult,
    WriteResult,
)

_READ_ONLY = "ChatStoreBackend доступен только для чтения (история чатов/проектов)"

BearerFn = Callable[[], str | None]


class ChatStoreBackend:
    """Read-only backend истории (deepagents ``CompositeBackend``-совместимый).

    Монтируется под якорь (напр. ``/history/``); parse толерантен к обрезке префикса
    ``CompositeBackend`` — якоря ищутся среди сегментов пути.
    """

    def __init__(
        self,
        *,
        base_url: str,
        bearer: BearerFn | None = None,
        http_client: httpx.AsyncClient | None = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._bearer = bearer or current_bearer
        self._client = http_client

    # ── ls ──────────────────────────────────────────────────────────────────────
    async def ls(self, path: str) -> LsResult:
        try:
            p = _parse_path(path)
            kind = p.kind
            if kind == "root":
                return LsResult(files=[_dir("/projects/"), _dir("/threads/")])
            if kind == "projects":
                data = await self._api("/api/projects/")
                return LsResult(files=[_dir(_project_dir(pr)) for pr in data.get("projects", [])])
            if kind == "project":
                return LsResult(files=[_dir(f"/projects/{p.project_slug}/threads/")])
            if kind == "project-threads":
                data = await self._api(f"/api/projects/{_enc(p.project_slug)}/threads/")
                return LsResult(
                    files=[
                        _thread_file(f"/projects/{p.project_slug}/threads/{_thread_seg(t)}", t)
                        for t in data.get("threads", [])
                    ]
                )
            if kind == "threads":
                data = await self._api("/api/threads/", {"projectId": "none"})
                return LsResult(
                    files=[
                        _thread_file(f"/threads/{_thread_seg(t)}", t)
                        for t in data.get("threads", [])
                    ]
                )
            return LsResult(error=f"Не директория: {path}")
        except Exception as exc:  # noqa: BLE001
            return LsResult(error=str(exc))

    # ── read ────────────────────────────────────────────────────────────────────
    async def read(
        self, file_path: str, offset: int | None = None, limit: int | None = None
    ) -> ReadResult:
        try:
            md = await self._render_read(_parse_path(file_path))
            if md is None:
                return ReadResult(error=f"Не найдено: {file_path}")
            return ReadResult(content=_slice_lines(md, offset, limit), mime_type="text/markdown")
        except Exception as exc:  # noqa: BLE001
            return ReadResult(error=str(exc))

    # ── glob (по имени, серверный ILIKE) ─────────────────────────────────────────
    async def glob(self, pattern: str, path: str | None = None) -> GlobResult:
        try:
            name = _strip_glob(pattern)
            p = _parse_path(path or "/")
            if p.kind in ("root", "projects"):
                data = await self._api("/api/projects/", {"name": name})
                return GlobResult(files=[_dir(_project_dir(pr)) for pr in data.get("projects", [])])
            if p.kind in ("project", "project-threads"):
                data = await self._api(
                    f"/api/projects/{_enc(p.project_slug)}/threads/", {"name": name}
                )
                return GlobResult(
                    files=[
                        _thread_file(f"/projects/{p.project_slug}/threads/{_thread_seg(t)}", t)
                        for t in data.get("threads", [])
                    ]
                )
            if p.kind == "threads":
                data = await self._api("/api/threads/", {"name": name, "projectId": "none"})
                return GlobResult(
                    files=[
                        _thread_file(f"/threads/{_thread_seg(t)}", t)
                        for t in data.get("threads", [])
                    ]
                )
            return GlobResult(files=[])
        except Exception as exc:  # noqa: BLE001
            return GlobResult(error=str(exc))

    # ── grep (по содержимому, серверный FTS) ─────────────────────────────────────
    async def grep(
        self, pattern: str, path: str | None = None, glob: str | None = None
    ) -> GrepResult:
        try:
            p = _parse_path(path or "/")
            if p.kind in ("project", "project-threads"):
                data = await self._api(
                    f"/api/projects/{_enc(p.project_slug)}/threads/", {"content": pattern}
                )
            elif p.kind == "threads":
                data = await self._api("/api/threads/", {"content": pattern, "projectId": "none"})
            else:
                data = await self._api("/api/projects/", {"content": pattern})
            matches = []
            for hit in data.get("matches", []):
                title = hit.get("threadTitle") or "чат"
                snippet = _one_line(str(hit.get("snippet", "")))
                matches.append(
                    GrepMatch(path=_grep_hit_path(hit), line=1, text=f"[{title}] {snippet}")
                )
            return GrepResult(matches=matches)
        except Exception as exc:  # noqa: BLE001
            return GrepResult(error=str(exc))

    # ── read-only ─────────────────────────────────────────────────────────────────
    async def write(self, file_path: str, content: str) -> WriteResult:
        return WriteResult(error=_READ_ONLY)

    async def edit(
        self, file_path: str, old_string: str, new_string: str, replace_all: bool = False
    ) -> EditResult:
        return EditResult(error=_READ_ONLY)

    async def read_raw(self, file_path: str) -> ReadRawResult:
        return ReadRawResult(error=_READ_ONLY)

    # ── рендереры read ──────────────────────────────────────────────────────────────
    async def _render_read(self, p: _Parsed) -> str | None:
        kind = p.kind
        if kind in ("root", "projects"):
            data = await self._api("/api/projects/")
            header = "# История" if kind == "root" else "# Проекты"
            lines = [header, ""]
            if kind == "root":
                lines.append("## Проекты")
            for pr in data.get("projects", []):
                lines.append(f"- {pr.get('title')} — `/projects/{_proj_seg(pr)}/`")
            if not data.get("projects"):
                lines.append("_нет проектов_")
            if kind == "root":
                lines += ["", "## Чаты вне проектов — `/threads/`"]
            return "\n".join(lines)
        if kind == "project":
            project = (await self._api(f"/api/projects/{_enc(p.project_slug)}/")).get("project", {})
            threads = (await self._api(f"/api/projects/{_enc(p.project_slug)}/threads/")).get(
                "threads", []
            )
            return _render_thread_list(
                f"# Проект: {project.get('title')}", threads, f"/projects/{p.project_slug}/threads"
            )
        if kind == "project-threads":
            threads = (await self._api(f"/api/projects/{_enc(p.project_slug)}/threads/")).get(
                "threads", []
            )
            return _render_thread_list(
                "# Чаты проекта", threads, f"/projects/{p.project_slug}/threads"
            )
        if kind == "threads":
            threads = (await self._api("/api/threads/", {"projectId": "none"})).get("threads", [])
            return _render_thread_list("# Чаты вне проектов", threads, "/threads")
        if kind == "project-thread":
            messages = (
                await self._api(
                    f"/api/projects/{_enc(p.project_slug)}/threads/{_enc(p.thread_slug)}"
                )
            ).get("messages", [])
            return _render_chat(p.thread_slug, messages)
        if kind == "thread":
            messages = (await self._api(f"/api/threads/{_enc(p.thread_slug)}")).get("messages", [])
            return _render_chat(p.thread_slug, messages)
        return None

    # ── HTTP ─────────────────────────────────────────────────────────────────────
    async def _api(self, path: str, query: dict[str, str] | None = None) -> dict[str, Any]:
        url = f"{self._base_url}{path}"
        headers = {"Accept": "application/json"}
        token = self._bearer()
        if token:
            headers["Authorization"] = f"Bearer {token}"
        params = {k: v for k, v in (query or {}).items() if v is not None}
        if self._client is not None:
            resp = await self._client.get(url, params=params, headers=headers)
        else:
            async with httpx.AsyncClient() as client:
                resp = await client.get(url, params=params, headers=headers)
        if resp.status_code >= 400:
            raise RuntimeError(f"chat-backend {path} -> HTTP {resp.status_code}")
        return resp.json()


# ── path parsing (толерантно к scope-обрезке: якоря ищутся в сегментах) ──────────────
@dataclass
class _Parsed:
    kind: str
    project_slug: str = ""
    thread_slug: str = ""


def _parse_path(path: str) -> _Parsed:
    seg = [s for s in path.split("/") if s]
    if not seg:
        return _Parsed("root")
    anchors = [i for i, s in enumerate(seg) if s in ("projects", "threads")]
    if not anchors:
        return _Parsed("unknown")
    i = anchors[0]
    if seg[i] == "projects":
        if i + 1 >= len(seg):
            return _Parsed("projects")
        project_slug = seg[i + 1]
        if i + 2 >= len(seg):
            return _Parsed("project", project_slug=project_slug)
        if seg[i + 2] == "threads":
            if i + 3 < len(seg):
                return _Parsed(
                    "project-thread", project_slug=project_slug, thread_slug=seg[i + 3]
                )
            return _Parsed("project-threads", project_slug=project_slug)
        return _Parsed("unknown")
    if i + 1 < len(seg):
        return _Parsed("thread", thread_slug=seg[i + 1])
    return _Parsed("threads")


# ── helpers ──────────────────────────────────────────────────────────────────────────
def _dir(path: str) -> FileInfo:
    return FileInfo(path=path, is_dir=True)


def _thread_file(path: str, t: dict[str, Any]) -> FileInfo:
    return FileInfo(path=path, is_dir=False, modified_at=t.get("updatedAt"))


def _proj_seg(p: dict[str, Any]) -> str:
    return p.get("slug") or p.get("id") or ""


def _project_dir(p: dict[str, Any]) -> str:
    return f"/projects/{_proj_seg(p)}/"


def _thread_seg(t: dict[str, Any]) -> str:
    return t.get("slug") or t.get("contextId") or ""


def _grep_hit_path(hit: dict[str, Any]) -> str:
    ts = hit.get("threadSlug") or hit.get("contextId") or ""
    project_slug = hit.get("projectSlug")
    return f"/projects/{project_slug}/threads/{ts}" if project_slug else f"/threads/{ts}"


def _render_thread_list(header: str, threads: list[dict[str, Any]], base: str) -> str:
    lines = [header, ""]
    for t in threads:
        title = t.get("title") or "Без названия"
        lines.append(f"- {title} — `{base}/{_thread_seg(t)}` (обновлён: {t.get('updatedAt')})")
    if not threads:
        lines.append("_нет чатов_")
    return "\n".join(lines)


def _render_chat(slug: str, messages: list[dict[str, Any]]) -> str:
    roles = {"user": "Пользователь", "assistant": "Ассистент"}
    lines = [f"# Чат {slug}", ""]
    for m in messages:
        role = roles.get(m.get("role", ""), m.get("role", ""))
        content = m.get("content") if isinstance(m.get("content"), str) else ""
        lines += [f"**{role}:** {content}", ""]
    if not messages:
        lines.append("_пусто_")
    return "\n".join(lines)


def _enc(s: str) -> str:
    return quote(s, safe="")


def _one_line(text: str) -> str:
    return re.sub(r"\s+", " ", text).strip()[:200]


def _strip_glob(pattern: str) -> str:
    """glob-маска → подстрока для серверного ILIKE (спецсимволы glob убираем)."""
    return re.sub(r"[*?]", "", pattern).strip()


def _slice_lines(content: str, offset: int | None, limit: int | None) -> str:
    if not offset and limit is None:
        return content
    lines = content.split("\n")
    start = offset or 0
    end = len(lines) if limit is None else start + limit
    return "\n".join(lines[start:end])
